runtime/src/main/scala/com/armanbilge/vilcacora/runtime/Interpreter.scala

```diff
@@ -0,0 +1,305 @@
+      param.degree = op.kernelParams.drop(2).headOption.map(_.toInt).getOrElse(3)
+
+      op.classLabels.zipWithIndex.foreach { case (l, i) => !(label + i) = l.toInt }
+      op.rho.zipWithIndex.foreach { case (r, i) => !(rho + i) = r }
+      op.vectorsPerClass.zipWithIndex.foreach { case (n, i) => !(nSV + i) = n.toInt }
+
+      svRows.zipWithIndex.foreach { case (svRowPtr, i) =>
+        !(svs + i) = svRowPtr
+        for (j <- 0 until numFeatures) {
+          val node = svRowPtr + j
+          node.index = j + 1
+          node.value = op.supportVectors(i * numFeatures + j)
+        }
+        (svRowPtr + numFeatures).index = -1
+      }
+
+      coefRows.zipWithIndex.foreach { case (coefRowPtr, i) =>
+        !(svCoefs + i) = coefRowPtr
+        for (j <- 0 until numSupportVectors) {
+          // ==========================================================
+
+          // ONNX coefficients are flat with shape [(nr_class-1), num_support_vectors].
+          // `i` is the class-pair index, `j` is the support-vector index.
+          // This formula correctly accesses the flat array in row-major order.
+          val index = i * numSupportVectors + j
+          !(coefRowPtr + j) = op.coefficients(index)
+          // ==========================================================
+        }
+      }
+
+      model.param = param
+      model.nr_class = nrClass
+      model.l = numSupportVectors
+      model.label = label
+      model.rho = rho
+      model.nSV = nSV
+      model.SV = svs
+      model.sv_coef = svCoefs
+      model
+    }
+  }
+
+  /** Copies input Scala Arrays into their corresponding pre-allocated C memory. */
+  private def copyInputsToMemory(
+      inputs: Map[String, Array[_]],
+      model: ModelIR,
+      memory: MemoryMap,
+  ): IO[Unit] = IO {
+    inputs.foreach { case (name, dataArray) =>
+      val ptr = memory.getOrElse(
+        name,
+        throw new NoSuchElementException(s"Input tensor '$name' not found in memory map"),
+      )
+      (dataArray, model.allocations(name).dataType) match {
+        case (data: Array[Float], DataType.Float32) =>
+          memcpy(ptr, data.at(0).asInstanceOf[Ptr[Byte]], data.length.toUSize * sizeof[CFloat])
+        case (data: Array[Double], DataType.Float64) =>
+          memcpy(ptr, data.at(0).asInstanceOf[Ptr[Byte]], data.length.toUSize * sizeof[CDouble])
+        case (data: Array[Int], DataType.Int32) =>
+          memcpy(ptr, data.at(0).asInstanceOf[Ptr[Byte]], data.length.toUSize * sizeof[CInt])
+        case (data: Array[Long], DataType.Int64) =>
+          memcpy(ptr, data.at(0).asInstanceOf[Ptr[Byte]], data.length.toUSize * sizeof[CLong])
+        case (_, dt) =>
+          throw new ClassCastException(s"Input data type mismatch for '$name', expected $dt")
+      }
+    }
+  }
+
+  /** Copies final results from C memory back into new Scala Arrays. */
+  private def copyOutputsFromMemory(model: ModelIR, memory: MemoryMap): IO[Map[String, Array[_]]] =
```

Comment:
Same thing here, instead of copying the memory back, we can directly write into the result array.